accurately resolve aliases on macOS

watters opened the pull request:
This change resolves the issue described in issue #47.

Recent versions of Firefox Developer Edition and Firefox Nightly use spaces in their app names—`Firefox Developer Edition.app` and `Firefox Nightly.app` respectively.

I tried to be sensitive to preserving some backwards compatibility with older names, in case folks are still using them. My local testing (which I acknowledge may miss a historical case I'm not aware of) succeeded in launching the correct version of Firefox when I renamed `/Applications/Firefox Nightly.app` -> `/Applications/FirefoxNightly.app` and `Firefox Developer Edition.app` -> `FirefoxDeveloperEdition.app`.

All of the tests supplied pass for me.

My manual testing included successfully running the following commands with the expected results:

```
./bin/fx-runner -b deved --foreground start
./bin/fx-runner -b firefoxdeveloperedition --foreground start
./bin/fx-runner -b nightly --foreground start
```

Happy to make any changes needed.

Rob--W commented on lib/utils.js:
```diff
@@ -45,6 +45,12 @@ function normalizeBinary (binaryPath, platform, arch) {
+
+        // recent versions (as of July 2020) of Firefox Developer Edition use
+        // 'org.mozilla.firefoxdeveloperedition' as the CFBundleName
+        if (!result && binaryPath === "deved") {
```
`deved` had never been an alias, it was recently introduced in #44.

Let's fix this specific part by exchanging "deved" and "firefoxdeveloperedition" in the implementation.

Rob--W commented on lib/utils.js:
```diff
@@ -114,9 +120,9 @@ function normalizeBinary (binaryPath, platform, arch) {
 normalizeBinary.paths = {
   "firefox on osx": "/Applications/Firefox.app/Contents/MacOS/firefox-bin",
   "beta on osx": "/Applications/FirefoxBeta.app/Contents/MacOS/firefox-bin",
```
Have you checked whether FirefoxBeta without spaces is still accurate?

watters replied:
When I download the Firefox Beta for macOS using the link at https://www.mozilla.org/en-US/firefox/channel/desktop/, the name of the downloaded application is simply `Firefox.app`, and the `CFBundleName` it is using is the same as the non-beta release:

```
        <key>CFBundleIdentifier</key>
        <string>org.mozilla.firefox</string>
```

As such, *I chose to left this unchanged just in case anyone is currently relying on the current implementation by manually naming downloaded beta versions `FirefoxBeta.app`* (I'm not familiar enough with the whole breadth of the Firefox build result output to know if there's another source for builds of the beta which may be differently named).

watters replied:
This behavior appears to parallel the behavior for Windows, which resolves the 'beta' and 'firefox' aliases to 'Mozilla Firefox', which is explained in a comment for resolving the Windows beta alias.

Happy to make the Mac resolution behave similarly to Windows.

Rob--W replied:
Firefox Release and Firefox Beta share the same installation locations on Windows and macOS.
The DeveloperEdition is Firefox Beta with a unique installation location.

And Aurora is defunct and dropped 3 years ago - https://release.mozilla.org/firefox/release/2017/04/17/dawn-project-faq

watters replied:
I've aligned the handling of the `beta` alias on macOS to the handling on Windows. It's isolated in a single commit, so easy to ditch if that's undesired.

Rob--W commented on test/run/test.utils.js:
```diff
@@ -230,24 +230,28 @@ describe("lib/utils", function () {
 
+    var defaultCanonicalPath = "/Applications/Firefox.app/Contents/MacOS/firefox-bin";
+    var devedCanonicalPath = "/Applications/Firefox Developer Edition.app/Contents/MacOS/firefox-bin";
+    var nightlyCanonicalPath = "/Applications/Firefox Nightly.app/Contents/MacOS/firefox-bin";
```
Let's just keep the strings inline. It's a bit repetitive, but it is easier to see the expectations by just looking at the lines.

watters commented on lib/utils.js:
```diff
@@ -25,9 +25,6 @@ function normalizeBinary (binaryPath, platform, arch) {
     platform = platform || os.platform();
     arch = arch || os.arch();
     binaryPath = binaryPath || process.env.JPM_FIREFOX_BINARY || "firefox";
-    if (binaryPath === "firefoxdeveloperedition") {
```
Sorry for being slow to understand your intention here—

When you say:

> This should not be removed, but swapped.

Do you mean that you want to accept `deved` as an alias to `fx-runner` (and, transitively, `web-ext`) which gets expanded as needed to `firefoxdeveloperedition` and drop support for `firefoxdeveloperedition`?

I'll commit a couple of options that try to do what I think you're asking.

watters replied:
Committed a change which adds back support for `deved` as an alias and preserving backward compatibility for `firefoxdeveloperedition`.

Happy to change this further. Sorry again for being slow to understand your desire/intent.

Rob--W commented:
This looks good, thanks. Could you undo the version bump? I would like to create a separate commit for that after merging.

I'm going to squash all commits and change the title to:

```
fix: accurately resolve aliases on macOS (#48)

Fixes #47
```

If you'd like, you can also squash the commits yourself when you update the PR, then I'll just merge the PR directly without squashing first.

watters commented:
Squashed commits, eliding the change to `package-lock.json`, and updated the commit message to align more closely with what you described above.

Thanks for your patience.

Rob--W commented:
Thanks for your patch!
